### PR TITLE
Port ekf_localizer

### PR DESCRIPTION
**Pull request by esteve**

### CODE REVIEW

**Comment by esteve** [comment]
@mitsudome-r I've been debugging this yesterday and today and I can't figure out why the test does not pass. Also, I've noticed that our CI is broken, since this PR is marked as passing, but tests don't actually pass (https://github.com/tier4/Pilot.Auto/pull/16/checks?check_run_id=1322754752). Do you or someone else have more expertise with `ekf_localizer` ? Thanks.

**Comment by mitsudome-r** [comment]
@esteve Tests for EKF is not working in ROS1, so please remove tests from CMakeLists.txt.

**Comment by esteve** [comment]
@mitsudome-r @TakaHoribe I've cleaned up the commit history and disabled the tests. The tests compile for ROS 2, but they don't pass. This PR is now ready for review. Thanks.

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp [hunk 1]
```diff
@@ -67,77 +74,91 @@ EKFLocalizer::EKFLocalizer() : nh_(""), pnh_("~"), dim_x_(6 /* x, y, yaw, yaw_bi
...
   sub_twist_with_cov_ =
-    nh_.subscribe("in_twist_with_covariance", 1, &EKFLocalizer::callbackTwistWithCovariance, this);
-  sub_twist_ = nh_.subscribe("in_twist", 1, &EKFLocalizer::callbackTwist, this);
+    create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>("in_twist_with_covariance", 1, std::bind(&EKFLocalizer::callbackTwistWithCovariance, this, _1));
+  sub_twist_ = create_subscription<geometry_msgs::msg::TwistWithCovarianceStamped>("in_twist", 1, std::bind(&EKFLocalizer::callbackTwist, this, _1));
 
   dim_x_ex_ = dim_x_ * extend_state_step_;
 
+  tf_br_ = std::make_shared<tf2_ros::TransformBroadcaster>(
+      std::shared_ptr<rclcpp::Node>(this, [](auto) {}));
```
Comment (TakaHoribe) [inline localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:114]: We can use `shared_from_this()` here, like this.

```
tf_br_ = std::make_shared<tf2_ros::TransformBroadcaster>(shared_from_this());
```

https://github.com/tier4/Pilot.Auto/pull/18#discussion_r506003731

File: localization/pose_twist_fusion_filter/ekf_localizer/CMakeLists.txt [hunk 2]
```diff
@@ -1,82 +1,41 @@
...
+#     src/kalman_filter/time_delay_kalman_filter.cpp
+#   )
+#   target_include_directories(ekf_localizer-test
+#     PRIVATE
+#     include
+#   )
+#   ament_target_dependencies(ekf_localizer-test geometry_msgs rclcpp tf2 tf2_ros)
+# endif()
+
+ament_auto_package()
```
Comment (TakaHoribe) [inline localization/pose_twist_fusion_filter/ekf_localizer/CMakeLists.txt:41]: Add install for launch in ament_auto_package().

```
ament_auto_package(
  INSTALL_TO_SHARE
  launch
)
```

**Review by TakaHoribe — COMMENTED** [review]
You need to modify the launch file.
I changed two things
 - `type="ekf_localizer"` to `exec="ekf_localizer"` when run node
 - `$(arg ***)` to `$(var ***)` for variables

then I could launch it.

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp (hunk 1 above)
Comment (TakaHoribe) [reply localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:114]: Sorry, I misunderstood. It is not used here.

File: localization/pose_twist_fusion_filter/ekf_localizer/include/ekf_localizer/ekf_localizer.h [hunk 3]
```diff
@@ -36,32 +36,29 @@
...
-  ros::Publisher pub_pose_cov_no_yawbias_;  //!< @brief ekf estimated yaw bias publisher
-  ros::Subscriber sub_initialpose_;         //!< @brief initial pose subscriber
-  ros::Subscriber sub_pose_;                //!< @brief measurement pose subscriber
-  ros::Subscriber sub_twist_;               //!< @brief measurement twist subscriber
-  ros::Subscriber sub_pose_with_cov_;       //!< @brief measurement pose with covariance subscriber
-  ros::Subscriber sub_twist_with_cov_;      //!< @brief measurement twist with covariance subscriber
-  ros::Timer timer_control_;                //!< @brief time for ekf calculation callback
-  ros::Timer timer_tf_;                     //!< @brief timer to send transform
-  tf2_ros::TransformBroadcaster tf_br_;     //!< @brief tf broadcaster
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::PoseStamped>> pub_pose_;             //!< @brief ekf estimated pose publisher
```
[the remark this answers is not in the document]
Comment (esteve) [reply localization/pose_twist_fusion_filter/ekf_localizer/include/ekf_localizer/ekf_localizer.h]: Done in 08dad92f4555a78570df6306b1e01309315da14f

File: localization/pose_twist_fusion_filter/ekf_localizer/include/ekf_localizer/ekf_localizer.h [hunk 4]
```diff
@@ -36,32 +36,29 @@
...
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>> pub_pose_cov_;         //!< @brief estimated ekf pose with covariance publisher
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::TwistStamped>> pub_twist_;            //!< @brief ekf estimated twist publisher
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::TwistWithCovarianceStamped>> pub_twist_cov_;        //!< @brief ekf estimated twist with covariance publisher
+  std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float64MultiArray>> pub_debug_;            //!< @brief debug info publisher
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::PoseStamped>> pub_measured_pose_;    //!< @brief debug measurement pose publisher
+  std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float64>> pub_yaw_bias_;         //!< @brief ekf estimated yaw bias publisher
```
[the remark this answers is not in the document]
Comment (esteve) [reply localization/pose_twist_fusion_filter/ekf_localizer/include/ekf_localizer/ekf_localizer.h]: Done in 8e1fe45260ae49fe82af7938cd63528aed6d81fe

File: localization/pose_twist_fusion_filter/ekf_localizer/CMakeLists.txt (hunk 2 above)
Comment (esteve) [reply localization/pose_twist_fusion_filter/ekf_localizer/CMakeLists.txt]: Done in 175194f6a91108aa26496982b56541802148d3d1

**Comment by esteve** [comment]
@TakaHoribe thanks for the review. Can you check if the launch file (bcbf4287949c9ac1038451ea431586c0c6391ccd) changes are correct? Thanks.

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp [hunk 5]
```diff
@@ -684,37 +705,37 @@ void EKFLocalizer::publishEstimateResult()
...
   twist_cov.twist.covariance[0] = P(IDX::VX, IDX::VX);
   twist_cov.twist.covariance[5] = P(IDX::VX, IDX::WZ);
   twist_cov.twist.covariance[30] = P(IDX::WZ, IDX::VX);
   twist_cov.twist.covariance[35] = P(IDX::WZ, IDX::WZ);
-  pub_twist_cov_.publish(twist_cov);
+  pub_twist_cov_->publish(twist_cov);
 
   /* publish yaw bias */
-  std_msgs::Float64 yawb;
+  autoware_debug_msgs::msg::Float64Stamped yawb;
```
Comment (TakaHoribe) [inline localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:729]: please add timestamp
```
yawb.stamp = current_time;
```

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp [hunk 6]
```diff
@@ -723,11 +744,11 @@ void EKFLocalizer::publishEstimateResult()
   if (current_pose_ptr_ != nullptr)
     pose_yaw = tf2::getYaw(current_pose_ptr_->pose.orientation) * RAD2DEG;
 
-  std_msgs::Float64MultiArray msg;
+  autoware_debug_msgs::msg::Float64MultiArrayStamped msg;
```
Comment (TakaHoribe) [inline localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:747]: please add timestamp
```
msg.stamp = current_time;
```

**Review by TakaHoribe — COMMENTED** [review]
Thank you. Almost looks good. Just one thing, please add timestamp with `current_time` for the `autoware_debug_msgs` values. After that, I'll merge this PR.

**Comment by TakaHoribe** [comment]
Also launch file looks good and works correctly.

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp (hunk 5 above)
Comment (esteve) [reply localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:729]: Added in 051586f

File: localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp (hunk 6 above)
Comment (esteve) [reply localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp:748]: Added in 051586f

**Review by TakaHoribe — APPROVED** [review]
LGTM.

**Pull request merged**